packages/@uppy/dashboard/src/components/FileList.js

```diff
@@ -64,7 +65,13 @@ module.exports = (props) => {
     handleCancelThumbnail: props.handleCancelThumbnail,
   }
 
-  const rows = chunks(Object.keys(props.files), props.itemsPerRow)
+  const sortByGhostComesFirst = (file1, file2) => {
+    return props.files[file2].isGhost - props.files[file1].isGhost
+  }
+
+  // Sort files by file.isGhost, ghost files first, only if recoveredState is present
+  const files = props.recoveredState ? Object.keys(props.files).sort(sortByGhostComesFirst) : Object.keys(props.files)
```

Comment:
`<array>.sort` mutates the array, you could simplify that bit (well at least to me it looks simpler, feel free to disagree):
```suggestion
  const files = Object.keys(props.files)
  if (props.recoveredState) files.sort(sortByGhostComesFirst)
```